engines/docker/sandbox.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/caching"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+const dockerEngineKillTimeout = 5 * time.Second
+
+type sandbox struct {
+	engines.SandboxBase
+	monitor     runtime.Monitor
+	containerID string
+	resultSet   engines.ResultSet
+	resultErr   error
+	abortErr    error
+	tempStorage runtime.TemporaryStorage
+	resolve     atomics.Once
+	client      *docker.Client
+	taskCtx     *runtime.TaskContext
+	handle      *caching.Handle
+}
+
+func newSandbox(sb *sandboxBuilder) (*sandbox, error) {
+	// create the container
+	opts := docker.CreateContainerOptions{
+		Config: sb.generateDockerConfig(),
+	}
+
+	container, err := sb.e.client.CreateContainer(opts)
+	if err != nil {
+		return nil, runtime.NewMalformedPayloadError(
+			"could not create container: " + err.Error())
+	}
+
+	// create a temporary storage for use by resultSet
+	ts, err := sb.e.Environment.TemporaryStorage.NewFolder()
+	if err != nil {
+		// unsure if this is the correct error type to return
+		return nil, runtime.NewMalformedPayloadError(
+			"could not create temporary storage")
+	}
+	sbox := &sandbox{
+		containerID: container.ID,
+		tempStorage: ts,
+		client:      sb.e.client,
+		taskCtx:     sb.taskCtx,
+		handle:      sb.handle,
+		monitor:     sb.monitor.WithPrefix("docker-sandbox").WithTag("taskID", sb.taskCtx.TaskID),
+	}
+
+	// attach to the container before starting so that we get all the logs
+	_, err = sbox.client.AttachToContainerNonBlocking(docker.AttachToContainerOptions{
+		Container:    container.ID,
+		OutputStream: ioext.WriteNopCloser(sbox.taskCtx.LogDrain()),
+		Logs:         true,
+		Stdout:       true,
+		Stderr:       true,
+		Stream:       true,
+	})
+
+	// HostConfig is ignored by the remote API and is only kept for
+	// backward compatibility.
+	err = sbox.client.StartContainer(sbox.containerID, &docker.HostConfig{})
+	if err != nil {
+		return nil, runtime.ErrFatalInternalError
+	}
+
+	go sbox.wait()
+
+	return sbox, nil
+}
+
+func (s *sandbox) WaitForResult() (engines.ResultSet, error) {
+	s.resolve.Wait()
+	return s.resultSet, s.resultErr
+}
+
+func (s *sandbox) wait() {
+	exitCode, err := s.client.WaitContainer(s.containerID)
+	s.resolve.Do(func() {
+		s.resultSet = newResultSet(exitCode == 0, s.containerID, s.client,
+			s.tempStorage, s.handle, s.monitor.WithPrefix("result-set"))
+		if err != nil {
+			s.resultSet = newResultSet(false, s.containerID, s.client, s.tempStorage,
+				s.handle, s.monitor.WithPrefix("result-set"))
+		}
+		s.abortErr = engines.ErrSandboxTerminated
+	})
+}
+
+func (s *sandbox) Kill() error {
+	s.resolve.Do(func() {
+		debug("Sandbox.Kill()")
+		// maybe will use to timeout kill and send SIGKILL
+		killCtx, cancelFunc := context.WithCancel(context.Background())
```

Comment:
See: https://golang.org/pkg/context/#WithCancel

>  Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this Context complete. 

Probably use [`WithTimeout`](https://golang.org/pkg/context/#WithTimeout), like:
```go
ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
defer cancel() 
```

----
Even better something like:

```go
ctx, cancel := context.WithTimeout(context.Background(), dockerEngineKillTimeout)
defer cancel()
_ = s.client.KillContainer(docker.KillContainerOptions{
  ID:      s.containerID,
  Signal:  docker.SIGTERM,
  Context: ctx,
})

err := s.client.KillContainer(docker.KillContainerOptions{
  ID:      s.containerID,
  Signal:  docker.SIGKILL,
})

if (err is NOT because container as already stopped) {
  monitor.reportError(err, 'KillContainer with SIGKILL failed')
}
```

Instead of checking whether the first call to `KillContainer` with `SIGTERM` was successful, we always call a second time with `SIGKILL`. If killing a stopped container causes an error we ignore that error, but otherwise we report any error that would happen in `SIGKILL`.

It's better to handle errors than to try avoid them happening... as you could have a race condition. Note: I don't know if killing a stopped container causes an error.